crates/stages/src/stages/index_account_history.rs

```diff
@@ -0,0 +1,472 @@
+                            *list.last().expect("Chuck does not return empty list") as TransitionId,
+                        ),
+                        TransitionList::new(list).expect("Indices are presorted and not empty"),
+                    )
+                })?;
+                // Insert last list with u64::MAX
+                if let Some(last_list) = last_chunk {
+                    tx.put::<tables::AccountHistory>(
+                        ShardedKey::new(address, u64::MAX),
+                        TransitionList::new(last_list)
+                            .expect("Indices are presorted and not empty"),
+                    )?
+                }
+                Ok(())
+            })?;
+
+        info!(target: "sync::stages::index_account_history", "Stage finished");
+        Ok(ExecOutput { stage_progress: to_block, done: true })
+    }
+
+    /// Unwind the stage.
+    async fn unwind(
+        &mut self,
+        tx: &mut Transaction<'_, DB>,
+        input: UnwindInput,
+    ) -> Result<UnwindOutput, StageError> {
+        info!(target: "sync::stages::index_account_history", to_block = input.unwind_to, "Unwinding");
+        let from_transition_rev = tx.get_block_transition(input.unwind_to)?;
+        let to_transition_rev = tx.get_block_transition(input.stage_progress)?;
+
+        let mut cursor = tx.cursor_write::<tables::AccountHistory>()?;
+
+        //<<DB as DatabaseGat>::TXMut as DbTxMutGAT>::CursorMut
+        tx.cursor_read::<tables::AccountChangeSet>()?
+            .walk(from_transition_rev)?
+            .take_while(|res| res.as_ref().map(|(k, _)| *k < to_transition_rev).unwrap_or_default())
+            .collect::<Result<Vec<_>, _>>()?
+            .into_iter()
+            // reverse so we can get lowest transition id where we need to unwind account.
+            .rev()
+            // fold all account and get last transition index
+            .fold(BTreeMap::new(), |mut accounts: BTreeMap<Address, u64>, (index, account)| {
+                // we just need address and lowest transition id.
+                accounts.insert(account.address, index);
+                accounts
+            })
+            .into_iter()
+            // try to unwind the index
+            .try_for_each(|(address, rem_index)| -> Result<(), StageError> {
+                let shard_part =
+                    unwind_account_history_shards::<DB>(&mut cursor, address, rem_index)?;
+
+                // check last shard_part, if present, items needs to be reinserted.
+                if !shard_part.is_empty() {
+                    // there are items in list
+                    tx.put::<tables::AccountHistory>(
+                        ShardedKey::new(address, u64::MAX),
+                        TransitionList::new(shard_part)
+                            .expect("There is at least one element in list and it is sorted."),
+                    )?;
+                }
+                Ok(())
+            })?;
```

Comment:
can we flatten this (maybe already commented this)? it's quite difficult to read. same in `IndexStorageHistoryStage`

Reply:
ok, left already comment like this. what i mean by flatten is 
```rust
// Collect all account changesets.
let account_changesets = tx.cursor_read::<tables::AccountChangeSet>()?
     // ... walk, take_while
     ;

// Create a map with lowest transition ids
let account_map = BTreeMap::default();
for entry in account_changesets {
    let (index, account) = entry?;
    account_map.insert(account.address, index); // maybe we need some logic to determine the lowest
}

for (address, rem_index) in account_map {
    // ... 
}
```

Reply:
wdym with flatten here, `collect` it and use `for _ in _`?

Reply:
It is flatten :). To be honest i still prefer iter chains, but this is nice too